src/handle_liquid.cpp

```diff
@@ -190,6 +190,14 @@ static bool get_liquid_target( item &liquid, const item *const source, const int
     }
 
     uilist menu;
+    menu.input_category = "LIQUID_TARGET";
+    input_context ctxt( "LIQUID_TARGET" );
+    menu.additional_actions = {
+        { "CONSUME", translation() },
+        { "CONTAINER_POUR", translation() },
+        { "GROUND_POUR", translation() }
+    };
```

Comment:
Do the hotkeys work in your test? From the look of it, the hotkeys will trigger the additional actions, which will be ignored by the `uilist` because they are not `ANY_INPUT` and `uilist::allow_additional` is not set.

And to be honest this seems a bit hacky to me. If you want a UI that handles actions instead of hotkeys you can use `query_popup`, but its layout is different from `uilist` so it might break consistency. Or we could upgrade `uilist` to handle actions instead of hotkeys, but that will be a large project.

Reply:
As noted in my testing, single letter hotkeys do work. Things like F10 do not show up. `uilist::allow_additional` is set on the next line. If this is just a misconfiguration of it being `ANY_INPUT`, do I just need to define the ctxt type (and possibly set the keybinding to be a specific input_method?

> And to be honest this seems a bit hacky to me.

I can't really disagree here because it definitely is, which is why I wanted your feedback to see if there's any easy ways to make it less hacky without having to upgrade `uilist`.

Reply:
> As noted in my testing, single letter hotkeys do work.

I my test when I press any of the three hotkeys, a message `never mind` is added to the message log and nothing else happens.

> Things like F10 do not show up

I think it's because `keys_bound_to` defaults to `restrict_to_printable=false` so the input event is not returned. `uilist` does not currently support hotkeys with a description longer than 2 either (which limits the hotkeys to a single character with at most one modifier key, or any other key with a short enough description such as F1).

> uilist::allow_additional is set on the next line

🤦‍♂️Can't believe I missed that, maybe I should get more sleep...

> If this is just a misconfiguration of it being ANY_INPUT, do I just need to define the ctxt type (and possibly set the keybinding to be a specific input_method?

You can handle the action (`uilist::ret_act`) if `uilist::ret` is `UILIST_ADDITIONAL` after `uilist::query` returns.

Here's another problem though: the player can change the keybindings by pressing `?` in the `uilist`, but the hotkeys will not be updated until they close and reopen the `uilist`.

Reply:
> I my test when I press any of the three hotkeys, a message never mind is added to the message log and nothing else happens.

I could have sworn this was working, but perhaps I also need to get more sleep...

> Here's another problem though: the player can change the keybindings by pressing ? in the uilist, but the hotkeys will not be updated until they close and reopen the uilist.

I'm pretty sure these updated immediately in my testing, but now you've got me second guessing things.



I'll take another look once your PR https://github.com/CleverRaven/Cataclysm-DDA/pull/57541 is merged. I won't be that sad if I can't make this work, but I figured I'd take a stab at it and see what was possible.  If it ends up causing more problems, I'm fine dropping it until a better solution can be implemented.